DKFZ-ODCF/FastqIndEx
# Add streamed data indexing

Add the possibility to stream in a FASTQ file for indexing. Internally,
I created the InputSource class for this, which is inherited by the
PathInputSource and the StreamInputSource classes. This way, both
classes share a common interface and can be treated equally. As we are
working with (char/byte)buffers, it is more convenient to have a C-like
interface with methods like read(), skip() and so on.

As the previous code was based on path instances, a lot of changes were
necessary to adapt to the new structure.

Note, that there are some tests, which still fail. They need to be
implemented.

Sources
=======

ActualRunner:
- Add additional constructors with a path and a shared pointer of type
  InputSource.
- Change the checkPremises method to be aware of the different input
  source types.
- Add allowsReadFromStreamedSource() which needs to be overriden, if
  necessary.

ExtractorRunner:
- Make it use a PathInputSource for input instead of a path.
NOTE: In difference to the indexer, the extractor is not allowed to
read from a stream (it would also not make any sense.)

IndexerRunner:
- Reformated the constructor code and added a second constructor.
- Override the allowsReadFromStreamedSource() method to return true.

Extractor:
- Add the resultFile, useFile and forceOverwrite fields. This allows the
  extractor to either write to stdout or to a result file.
- Extend the constructor to be aware of the new fields.
- Extend checkPremises to be aware of the changes.
- Adapt the extractor to use the new (Path)InputSource instead of path.

Indexer:
- Make the Indexer use an InputSource instance instead of a path.
- Make the indexing process use the new InputSource methods.

Starter:
- The Starter will now recognize "-" as a valid parameter for a piped
  FASTQ file in index mode and for the extracted text in extract mode.

Added the InputSource, PathInputSource and StreamInputSource classes.

Tests
=====

Made the test classes use the new Indexer/Extractor(Runner) constructors

Added tests for the new InputSource classes and updated the Indexer
tests to also test with streamed data.

Replaced several duplicate strings by constants.

Disabled the integration tests for now. Streamed data can be tested with
e.g. ifstream instances.

TestResourcesAndFunctions:
- Add a format string method. However, this method is not very safe and
  really only for test purposes. It should not be used in production
  code! It is ok for tests though.
- Add the runCommand(), runIndexerBinary(), runExtractorBinary(),
  extractGZFile() and createConcatenatedFile() methods to remove code
  duplication and simplify tests.
